File: mercury/pr005.m
:- module pr005.

% Problem 5
%
% 30 November 2001
%
% 2520 is the smallest number that can be divided by each of the numbers
% from 1 to 10 without any remainder.
%
% What is the smallest positive number that is evenly divisible by all of
% the numbers from 1 to 20?
%
% 232792560

:- interface.
:- import_module io.
:- pred main(io::di, io::uo) is det.

:- implementation.
:- import_module list, int, string.

main(!IO) :-
    Answer = int.fold_up(lcm, 1, 20, 1),
    io.format("%d\n", [i(Answer)], !IO).

:- func lcm(int, int) = int.
lcm(A, B) = (A div gcd(A, B)) * B.

:- func gcd(int, int) = (int) is det.
gcd(A, B) = C :-
    (if B = 0 then C = A else C = gcd(B, A mod B)).

File: nimrod/pr005.nim
# Problem 5
#
# 30 November 2001
#
#
# 2520 is the smallest number that can be divided by each of the numbers
# from 1 to 10 without any remainder.
#
# What is the smallest positive number that is evenly divisible by all of
# the numbers from 1 to 20?
#
# 232792560

proc gcd(a, b: int): int =
  var a = a
  var b = b
  var r = a mod b
  while r > 0:
    a = b
    b = r
    r = a mod b
  result = b

proc lcm(a, b: int): int =
  (a div gcd(a, b)) * b

var reduct = 1
for i in 2..20:
  reduct = lcm(reduct, i)
echo(reduct)
